### PR TITLE
Autoplanner will take number of pilots into account when planning purchases

### LINKED ISSUE #1832: Autopurchaser should not buy aircraft when there aren't enough pilots to fly them
**Is your feature request related to a problem? Please describe.**
At the moment, the autoplanner will purchase aircraft that it can task, without checking if there are enough pilots to fly them.

**Describe the solution you'd like**
Autoplanner should take pilot count into account when purchasing aircraft.

### PR DESCRIPTION
**Pull request by MetalStormGhost**
Added a check to make sure the number of aircraft won't exceed the number of available pilots after fulfilling autoplanner aircraft requests.

Resolves #1832

### CODE REVIEW

File: game/squadrons/airwing.py [hunk 1]
```diff
@@ -57,9 +58,15 @@ def best_squadrons_for(
...
-        return None
+        best_squadrons = self.best_squadrons_for(location, task, size, this_turn)
+        if not best_squadrons:
+            return None
+        else:
+            # In case we have several best squadrons at this point,
+            # return a random squadron from the list of best squadrons for this MissionTarget
+            # so the autoplanner will reserve pilots and aircraft from several squadrons,
+            # not just the first one in the list.
+            return random.choice(best_squadrons)
```
Comment (DanAlbert) [inline game/squadrons/airwing.py]: What does this have to do with the rest of the PR?

File: game/ato/flight.py [hunk 2]
```diff
@@ -122,6 +122,8 @@ def missing_pilots(self) -> int:
     def return_pilots_and_aircraft(self) -> None:
         self.roster.clear()
         self.squadron.claim_inventory(-self.count)
+        # Also release any pilots reserved by the autoplanner
+        self.squadron.release_autoplanner_pilot_reservations()
```
Comment (DanAlbert) [inline game/ato/flight.py]: Why do planned flights managing unrelated pilot reservations? Why does returning _one_ flight cancel _all_ reservations in the squadron? This needs a comment explaining why this needs to happen here.

File: game/squadrons/squadron.py [hunk 3]
```diff
@@ -43,6 +43,15 @@ class Squadron:
         default_factory=list, init=False, hash=False, compare=False
     )
 
+    #: A separate list of pilots for the autoplanner.
+    #: The pilots will be reserved from here during each turn when running the autoplanner,
+    #: before actual aircraft orders are made. This will prevent the autoplanner from making
+    #: several overlapping orders for the same squadron, potentially causing the number of
+    #: ordered aircraft from exceeding the number of available pilots.
+    autoplannable_pilots: list[Pilot] = field(
```
Comment (DanAlbert) [inline game/squadrons/squadron.py]: How does this differ in behavior from `available_pilots`? This sounds like the same thing. The _name_ is definitely describing the same thing. "Available" means "available to the auto planner" (and the player, but since the auto planner runs before the user _can_ interact, that's not a meaningful difference).

This implementation seems way overcomplicated to me. I think all we need is to check (during procurement) that the number of expected aircraft next turn does not exceed the number of expected pilots next turn. What am I missing that makes that not work?

File: game/procurement.py [hunk 4]
```diff
@@ -185,11 +189,21 @@ def fulfill_aircraft_request(
     ) -> Tuple[float, bool]:
         for squadron in squadrons:
             price = squadron.aircraft.price * quantity
+
+            # Final check to make sure the number of aircraft won't exceed the number of available pilots
+            # after fulfilling this aircraft request.
+            if (
+                squadron.pilot_limits_enabled
+                and squadron.owned_aircraft + squadron.pending_deliveries + quantity
```
Comment (DanAlbert) [inline game/procurement.py]: `squadron.expected_size_next_turn + quantity`

**Review by DanAlbert — CHANGES_REQUESTED** [review]

File: game/squadrons/airwing.py (hunk 1 above)
Comment (MetalStormGhost) [reply game/squadrons/airwing.py]: Currently, the autoplanner always seems to select the same squadron every time if the best_squadrons_for list has the same contents. This leads it to make purchases which are not evenly distributed across e.g. other squadrons operating out of the same base, with the same aircraft type and with identical task profiles:
image

In the above screenshot, there have been no combat losses. The differences in the aircraft counts are purely due to the best squadron selection and the procurement AI buying aircraft based on it.

My concern was that if I limit the procurement of aircraft to the max pilot limit, substituting purchases would simply be scrubbed instead of buying the aircraft for other capable squadrons, without this added random selection.

File: game/squadrons/squadron.py (hunk 3 above)
Comment (MetalStormGhost) [reply game/squadrons/squadron.py]: The key difference is that if the reservations are made during the autoplanner mission proposal stage, which means the autoplanner will select a different squadron if all pilots are already reserved from the first one. That will not happen if the impose the limit only at the budget spending stage, because substitute squadrons will not be looked for (during this turn, anyway) and the purchases would simply be scrubbed.

My understanding about available_pilots was that it's also used outside the autoplanner stage, which is why I took that as the data source for the pilot reservation list but implemented a separate data structure. It can, of course, be refactored but I would need to study how the available pilot handling works outside the autoplanner stage.

File: game/squadrons/airwing.py (hunk 1 above)
Comment (DanAlbert) [reply game/squadrons/airwing.py]: That's the intent. The auto planner is supposed to make good decisions, not random ones.

File: game/squadrons/squadron.py (hunk 3 above)
Comment (DanAlbert) [reply game/squadrons/squadron.py]: > That will not happen if the impose the limit only at the budget spending stage, because substitute squadrons will not be looked for (during this turn, anyway) and the purchases would simply be scrubbed.

That seems like something that should be fixed in the procurer though. When purchases happen pilots are already assigned, so the procurer has all the information it needs to know that a squadron shouldn't be considered and that it should move on to one that does have pilots available.

Comment (MetalStormGhost) [reply game/squadrons/squadron.py]: I'll take a look at refactoring this with the actual check for pilots done in the procurer. My hunch is that it will actually be a lot simpler that way. I'll probably get around to do that either tomorror or the day after.

File: game/procurement.py [hunk 5]
```diff
@@ -182,26 +182,49 @@ def most_needed_unit_class(self, cp: ControlPoint) -> GroundUnitClass:
...
+            insufficient_pilots = False
             price = squadron.aircraft.price * quantity
+
+            # Final check to make sure the number of aircraft won't exceed the number of available pilots
+            # after fulfilling this aircraft request.
+            if (
+                squadron.pilot_limits_enabled
+                and squadron.expected_size_next_turn + quantity > squadron.max_size
+            ):
+                insufficient_pilots = True
```
Comment (DanAlbert) [inline game/procurement.py]: For this to be correct it needs to happen _after_ the budget check. Right now this will return `True` for `insufficient_pilots` if _any_ squadron didn't have sufficient pilots. You only want to do that if _no_ squadron has enough pilots, otherwise you fall into the bad case described on L229.

File: game/procurement.py [hunk 6]
```diff
@@ -182,26 +182,49 @@ def most_needed_unit_class(self, cp: ControlPoint) -> GroundUnitClass:
...
                 # No airbases in range of this request. Skip it.
                 continue
-            budget, fulfilled = self.fulfill_aircraft_request(
+            budget, fulfilled, insufficient_pilots = self.fulfill_aircraft_request(
                 squadrons, request.number, budget
             )
+            if insufficient_pilots:
+                # The request was not fulfilled because there were not enough pilots in any of the provided squadrons
+                # and not because of insufficient budget. Continue processing the aircraft requests.
+                continue
```
Comment (DanAlbert) [inline game/procurement.py]: Won't this cause a similar problem as described on L229? These requests are in a priority order, and if we can't fulfill an early item then we probably can't plan the mission the later items were intended for either way.

If the orders here are something like:

2xDEAD
2xSEAD Escort
2xEscort
2xDEAD
2xSEAD Escort
2xEscort
2xDEAD
2xSEAD Escort
2xEscort

If there are only 6 parking spaces available, plenty of budget, and plenty of pilots for escort missions but none for SEAD Escort/DEAD, you'll end up filling the airport with escorts that can never be used because there will never be fulfilled DEAD missions for them to escort.

**Review by DanAlbert — CHANGES_REQUESTED** [review]

File: game/procurement.py (hunk 6 above)
Comment (MetalStormGhost) [reply game/procurement.py]: That is a possibility. I haven't encountered such cases yet in any of my campaigns, but I tend to design my own campaigns and my squadron configuration might work around the problem.

Escorts can (in most cases) be used for BARCAP, so the situation might resolve itself when some of them get shot down?

I think there needs to be some way to distinguish between purchase orders scrapped because of insufficient funds and orders scrubbed because of insufficient pilots. Otherwise the AI will almost stop buying aircraft altogether when squadrons get low on pilots.

Comment (MetalStormGhost) [reply game/procurement.py]: One possible solution would be to use hierachy in the aircraft procurement requests. Instead of putting the escorts in a flat list with the SEAD/DEAD, the escorts would be included in the same request. The we could scrub the whole mission including the escorts in case one of the squadrons runs out of pilots?

Comment (DanAlbert) [reply game/procurement.py]: It's not something you see for bluefor because most aircraft are multirole. It absolutely wrecks red factions because they end up buying hundreds of MiG-19s and nothing else and can then never plan any missions.

Purchases are in the order that missions are planned, and missions are often dependent on earlier missions. If you miss the first purchase but manage to fill every other purchase, there's no guarantee that _any_ additional missions will be planned.

I don't think there's a meaningful difference in why the purchase failed. Without explicit dependency tracking for packages and purchase orders, we cannot continue after an unfulfilled purchase regardless of why it happened or we'll cause the problem described by that comment.

Comment (MetalStormGhost) [reply game/procurement.py]: How about I make it an option (default: off) in the autoplanner settings under Campaign Management? By default, the first scrubbed purchase would cancel all the rest and the user could separately enable a setting to have the autoplanner continue processing subsequent orders if a squadron with insuffient pilots is encountered.

Comment (DanAlbert) [reply game/procurement.py]: I don't think a "buy aircraft that can't be used and potentially fill all the parking so it breaks my game" checkbox is something we should add. This isn't a theoretical problem. This is why a ton of factions couldn't sortie any missions but CAP several releases ago. Players will not notice that kind of failure when it happens to red because the only symptom is that the game is boring and easy.

File: game/procurement.py [hunk 7]
```diff
@@ -182,24 +182,43 @@ def most_needed_unit_class(self, cp: ControlPoint) -> GroundUnitClass:
...
+        return budget, False, insufficient_pilots
 
     def purchase_aircraft(self, budget: float) -> float:
         for request in self.game.coalition_for(self.is_player).procurement_requests:
             squadrons = list(self.best_squadrons_for(request))
             if not squadrons:
                 # No airbases in range of this request. Skip it.
                 continue
-            budget, fulfilled = self.fulfill_aircraft_request(
+            budget, fulfilled, insufficient_pilots = self.fulfill_aircraft_request(
```
Comment (DanAlbert) [inline game/procurement.py]: `insufficient_pilots` is unused. You can undo that change to the API and clean up the code that enables it.

**Review by DanAlbert — CHANGES_REQUESTED** [review]

File: game/procurement.py [hunk 8]
```diff
@@ -185,6 +185,13 @@ def fulfill_aircraft_request(
     ) -> Tuple[float, bool]:
         for squadron in squadrons:
             price = squadron.aircraft.price * quantity
+            # Final check to make sure the number of aircraft won't exceed the number of available pilots
+            # after fulfilling this aircraft request.
+            if (
+                squadron.pilot_limits_enabled
+                and squadron.expected_size_next_turn + quantity > squadron.max_size
```
Comment (DanAlbert) [inline game/procurement.py]: And now with the clarity of a much smaller PR, I see the bug :) `squadron.max_size` is the maximum number of pilots. You want the number of pilots that are expected to be present next turn. Add this to `Squadron` and use that:

```python
@property
def expected_pilots_next_turn(self) -> int:
    return len(self.active_pilots) + self.settings.squadron_replenishment_rate
```

**Review by DanAlbert — CHANGES_REQUESTED** [review]

File: game/procurement.py (hunk 8 above)
Comment (MetalStormGhost) [reply game/procurement.py]: This is now done, marking as ready for review.

**Comment by RndName** [comment]
LGTM, @DanAlbert i think we can merge it right away?

File: game/squadrons/squadron.py [hunk 9]
```diff
@@ -214,6 +214,10 @@ def _pilots_without_status(self, status: PilotStatus) -> list[Pilot]:
     def max_size(self) -> int:
         return self.settings.squadron_pilot_limit
 
+    @property
+    def expected_pilots_next_turn(self) -> int:
+        return len(self.active_pilots) + self.settings.squadron_replenishment_rate
```
Comment (RndName) [inline game/squadrons/squadron.py]: i think this has to be sth like
``` python
@property
 def expected_pilots_next_turn(self) -> int:
     return len(self.active_pilots) + self.replenish_rate
```
which uses the `self._number_of_unfilled_pilot_slots`

```python
@property
def replenish_rate(self) -> int:
    return min(self.settings.squadron_replenishment_rate, self._number_of_unfilled_pilot_slots)
```

otherwise the `expected_pilots_next_turn` will return more pilots than there actually would be (replenish rate > pilots required)

Comment (DanAlbert) [reply game/squadrons/squadron.py]: Good catch :)

Comment (MetalStormGhost) [reply game/squadrons/squadron.py]: Thanks for noticing this! Has been modified, as suggested. :)
